### PR TITLE
pkg/trace/{agent,api,writer}: improve queueing and receiver strategy.

### PR DESCRIPTION
**Pull request by knusbaum**
This PR changes how we approach queueing.

The approach in the agent so far has been to accept payloads as fast as possible and buffer them, never dropping anything and also never slowing down receiving payloads except by the rate limiter, controlled by the memory/cpu watchdog.

The rate limiter has not been effective at controlling memory footprint, since it is very slow to react. While we may be able to tune it to behave better, it is fundamentally the wrong approach for controlling memory usage in the agent.

This new approach is to use channel backpressure to limit the amount of in-flight payloads in the agent. Channels are limited to a much smaller number of items (often 1) rather than the very large channels we had previously (often 1000). Any time these channels would begin to fill up, we were in trouble with memory. What's worse is since the channels are likely never full, the receiver can continue to happily read as many payloads as it can, as fast as it can, increasing the memory load as well as using up cpu resources for decoding payloads that should instead be devoted to processing the payloads we've already accepted, as well as for GC.

Now, when the processing pipeline is backed up, the http receiver cannot send payloads on the channel, and the receiver blocks. There is a new limiter in the listeners that keeps the connection count low to prevent an explosion of goroutines, and the receiver itself is limited in the number of payloads it can decode simultaneously by a semaphore.

This means that senders receive backpressure. Submission of traces to the agent when it is under extreme load used to be as fast as when it was under no load. Now, trace submission time might increase, and in the worst case it will timeout or (ideally) receive a 429 response. The agent will still return a `200 OK` response by default, even when rejecting payloads, however a Tracer submitting payloads may include the `Datadog-Send-Real-Http-Status` header, set to any value, to signal to the agent to return a "real" response. In the case of rejected payloads, this means a `429 Too Many Requests` status.

In my experiments, the results were good with a surprisingly small number of simultaneous readers, likely because the agent is CPU-bound in the processing stage. Having a small number of concurrent decodings makes sure we aren't devoting too much processing power (and memory) to ingesting payloads.

### Possible Drawbacks / Trade-offs

This is a big change to the agent's strategy. We have to test it carefully to make sure the performance characteristics are good in general and not just for our synthetic tests.

### Describe how to test/QA your changes

We want to run this on the performance environment, relenv, and deploy to staging to see how this performs.
I've already performed manual tests with good results.

Update:
Extensive testing has been done by multiple teams. This includes:
* Local stress testing by APM Go team
* Testing by Java Tracer team, looking for anomalous or otherwise unacceptable behavior
* Performance benchmarks by the APM Go team in the team's performance testing environment
* Performance benchmarks by R&P team in the reliability environment
* Deployment of a build onto our staging cluster (affecting multiple backend teams) running since Aug. 03

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/trace/api/api.go
Comment (knusbaum) [inline pkg/trace/api/api.go:476]: Need to revert this.
Apparently we need to read the body into discard for the keep-alive to work.

File: pkg/trace/writer/trace.go
Comment (knusbaum) [inline pkg/trace/writer/trace.go:259]: Previously, launching a goroutine here meant that the memory consumption could grow unbounded.
If the sender channel was full the goroutines would just block waiting for room while new goroutines were launched.

File: pkg/trace/writer/trace.go
Comment (knusbaum) [inline pkg/trace/writer/trace.go:120]: This may be too low for climit. We need to test.

My local testing used a fake intake, so it was probably fast enough to keep up with whatever I sent to the agent.

File: pkg/trace/api/listener.go
Comment (knusbaum) [inline pkg/trace/api/listener.go:42]: 1000 connections should be more than enough to avoid connection refused in most reasonable cases. This just avoids the odd case where a connection explosion could cause unbounded goroutines to be spawned.

The actual limit is imposed in pkg/agent/api/api.go where we deserialize the payloads.

File: cmd/trace-agent/config/config.go
Comment (knusbaum) [inline cmd/trace-agent/config/config.go:254]: I may remove this, or at least keep it undocumented.

I haven't been able to get it to improve submission latency, throughput, or anything with simulated bursty traffic or any other kind of load.

There *might* be use-cases for it, but it is kind of a footgun.

**Comment by knusbaum** [comment]
Please see the release notes for more user-facing details.

File: pkg/trace/api/api.go
Comment (ajgajg1134) [inline pkg/trace/api/api.go:464]: Do we want to change this? I remember us chatting briefly about how some old tracers might rely on the agent to _not_ send a 429 response (Although I think if we _can_ move to 429 instead of some weird config option we should definitely do so)

File: pkg/trace/pb/agent_payload_vtproto.pb.go
Comment (ajgajg1134) [inline pkg/trace/pb/agent_payload_vtproto.pb.go:1]: Was this file changed intentionally? (among all the other pb files and `generate.sh`)?

File: pkg/trace/writer/sender.go
Comment (ajgajg1134) [inline pkg/trace/writer/sender.go:261]: Where did this `4` come from?

File: pkg/trace/writer/trace.go
Comment (ajgajg1134) [inline pkg/trace/writer/trace.go:106]: Where did this 100 come from?

**Review by ajgajg1134 — CHANGES_REQUESTED** [review]

File: pkg/trace/api/api.go
Comment (knusbaum) [reply pkg/trace/api/api.go:464]: My thinking is that we should push for this. We *may* have to go back and use the config, but I think enough time has passed from the initial implementation that we should at least revisit and heavily push that this is something tracers need to handle.

Definitely needs discussion and buy-in from the tracer teams though.

File: pkg/trace/pb/agent_payload_vtproto.pb.go
Comment (knusbaum) [reply pkg/trace/pb/agent_payload_vtproto.pb.go:1]: No, this must be leftover stuff. I'll revert it.

File: pkg/trace/writer/trace.go
Comment (knusbaum) [reply pkg/trace/writer/trace.go:106]: This was arrived at through experimentation. We can leave it at 200 if we want and I can retest.

I wanted to bring this down, the argument being that submitting payloads is inherently network-bound, not CPU bound, so the benefits of additional connections falls off precipitously, resulting only in additional RAM usage. The bandwidth is only so high and throwing concurrency at it will not help. 

However, I didn't want to be too aggressive. Cutting the former limit in half seemed like a good first step and with this configuration I got good results.

File: pkg/trace/writer/sender.go
Comment (knusbaum) [reply pkg/trace/writer/sender.go:261]: This is a magic number I picked.

It seems like 4 failed attemts is enough to say that the payload is not going to succeed. However definitely open to discussing what this number *should* be.

I'm not sure how this could/should be parameterized. For instance if the trace agent isn't under any load, we could hang onto payloads indefinitely. However, I'm not sure how valuable that work is relative to the complexity involved. Given the trace agent is usually not just sitting there doing nothing *while also* having payloads waiting to go out , my guess is it's not very valuable to try and adjust the retry count based on internal state.

Happy to hear thoughts.

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:13]:
```suggestion
    backpressure in the trace Agent, causing it to reject payloads when it
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:19]:
```suggestion
    be adjusted, as the defaults have been tested and confirmed to be sufficient
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:24]:
```suggestion
      This controls the size of the buffer that holds traces which were accepted
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:26]:
```suggestion
      be adjusted, as it does not increase throughput, but increases the RAM usage
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:27]:
```suggestion
      of the trace Agent considerably. The only time when this may be useful is for
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:32]:
```suggestion
      payloads. By default, this is max(1, 0.5 * (# cores allowed to the trace Agent)),
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:31]:
```suggestion
      This controls the number of simultaneous goroutines that are allowed to decode
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:37]:
```suggestion
      This controls how long decoder wait for a turn to decode a payload. There
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:42]:
```suggestion
      timeout is 1 second, after which the trace Agent sends a 429 Too Many Requests error.
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:46]:
```suggestion
      into the Agent's HTTP server. This number defaults to 1000, which is high enough that
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:47]:
```suggestion
      it should not take effect except for Agents with extreme numbers of clients.
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:48]:
```suggestion
      In practice, it's not expected to see this limit reached and it is only there
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:49]:
```suggestion
      to prevent accidental or intentional denial-of-service attacks on the trace Agent.
```

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:45]:
```suggestion
      This controls the maximum number of active network connections that are allowed
```

**Review by buraizu — APPROVED** [review]
Thanks, approving with some minor updates per the [documentation style guide](https://github.com/DataDog/documentation/blob/master/CONTRIBUTING.md)

**Comment by knusbaum** [comment]
Thanks, @buraizu. Would you take another look? I believe I addressed all your concerns and made a couple other clarifications.

File: pkg/trace/pb/agent_payload_vtproto.pb.go
Comment (knusbaum) [reply pkg/trace/pb/agent_payload_vtproto.pb.go:1]: Fixed.

File: pkg/trace/agent/agent.go
Comment (knusbaum) [inline pkg/trace/agent/agent.go:89]: TODO: reformat this log.

**Comment by dougqh** [comment]
> The approach in the agent so far has been to accept payloads as fast as possible and buffer them, never dropping anything and also never slowing down receiving payloads except by the rate limiter, controlled by the memory/cpu watchdog.

And in effect, this didn't really stop the trace agent from dropping, it just shifted the dropping to the tracer.  What we've seen with high traffic loads in resource constrained environments is that the trace agent becomes unresponsive.  Typically, this is from CPU throttling, but in theory, it could also be from a long GC pause.

When the trace agent is throttled and cannot keep up with the tracer, the tracer is forced to drop.  In short, the idea of not dropping traces didn't ever really work.  I think this is going to be a significant improvement.

> The rate limiter has not been effective at controlling memory footprint, since it is very slow to react. While we may be able to tune it to behave better, it is fundamentally the wrong approach for controlling memory usage in the agent.

I worked on similar system with others before and it was devilishly hard to get right.We're effectively trying to solve an allocation demand problem and that only way to do that is to keep some contingency space that you almost never use.

I think we're much better off designing something simple and resilient and just making sure we have health metrics to report how everything is working.

File: pkg/trace/api/api.go
Comment (dougqh) [inline pkg/trace/api/api.go:527]: Yikes, I can definitely see how this caused problems.

File: pkg/trace/api/api_test.go
Comment (dougqh) [inline pkg/trace/api/api_test.go:780]: I know this code is going away, but what???
Not reading a full body isn't a server bug...
I'm guessing this was because the old scheme required making sure the full payload was consumed and in-memory.

File: pkg/trace/api/api.go
Comment (knusbaum) [reply pkg/trace/api/api.go:527]: Exactly.

File: pkg/trace/api/api_test.go
Comment (knusbaum) [reply pkg/trace/api/api_test.go:780]: I think this was only used in tests to try and trigger some (unfortunately undescribed) bad behavior. The related test that used this is gone because the rate limiter is gone, so we don't need it. I can't figure out what exactly it was trying to test since, as you say, not reading a full body isn't a bug.

**Review by buraizu — APPROVED** [review]
Thanks for implementing those changes, LGTM

**Review by ajgajg1134 — APPROVED** [review]

File: pkg/trace/writer/trace.go
Comment (knusbaum) [reply pkg/trace/writer/trace.go:106]: Going to leave this out for now. We can add a follow-on if we discover we need it.

File: cmd/trace-agent/run.go
Comment (ahmed-mez) [inline cmd/trace-agent/run.go:257]: Not sure why we'd do this. Calling `agentrt.SetGoMemLimit` at L240 should be enough as it sets gomemlimit to 90% of the cgroup limit. Am I missing something?

File: pkg/trace/api/api.go
Comment (ahmed-mez) [inline pkg/trace/api/api.go:630]: I guess we should consider removing watchdog code and options entirely, any reason why we wouldn't do it in this PR?

File: pkg/trace/config/config.go
Comment (ahmed-mez) [inline pkg/trace/config/config.go:355]: I believe it's the number of concurrent decoding workers, not traces.

File: pkg/trace/writer/sender.go
Comment (ahmed-mez) [inline pkg/trace/writer/sender.go:230]: Is there a good reason to set it to 4? Should we also make it configurable?

**Review by ahmed-mez — COMMENTED** [review]
Thanks for working on this, a great improvements!
Added a few comments but need to review again the changes in the writer (my comments shouldn't block merging the PR).

File: pkg/trace/api/api.go
Comment (ahmed-mez) [inline pkg/trace/api/api.go:683]: We'll need to update https://docs.datadoghq.com/tracing/troubleshooting/agent_apm_metrics/

File: pkg/trace/config/config.go
Comment (knusbaum) [reply pkg/trace/config/config.go:355]: I think that's the same thing, but I will clarify.

File: pkg/trace/writer/sender.go
Comment (knusbaum) [reply pkg/trace/writer/sender.go:230]: I'm on the fence about this one. 4 is just a magic number chosen basically out of nowhere, so I'm happy to change it. 

OTOH, I would like to avoid adding more configs. This probably has an optimal value or range of values, and I don't think users should need to change it. The fewer knobs for users to turn, the better, not because we don't want them to have options, but because we don't want them to *have* to tune every parameter to get good behavior.

If we're not happy with 4, maybe we can experiment or discuss and find a better value.

File: pkg/trace/api/api.go
Comment (ahmed-mez) [reply pkg/trace/api/api.go:630]: +1 on removing the self-OOM later!

File: pkg/trace/config/config.go
Comment (ahmed-mez) [reply pkg/trace/config/config.go:355]: Couldn't find a better wording lol, let's keep it decoders 🤷

File: pkg/trace/writer/sender.go
Comment (ahmed-mez) [reply pkg/trace/writer/sender.go:230]: Having the option to change the value can be helpful to mitigate regressions without having to release a patch. That's the main reason behind the suggestion.

Comment (knusbaum) [reply pkg/trace/writer/sender.go:230]: :+1: That makes sense. Maybe we can have a config but keep it undocumented for now, perhaps removing it later. WDYT?

File: cmd/trace-agent/run.go
Comment (knusbaum) [reply cmd/trace-agent/run.go]: Fixed.

File: pkg/trace/api/api.go
Comment (knusbaum) [reply pkg/trace/api/api.go]: This has been moved into an opt-in header.

**Comment by knusbaum** [comment]
@ahmed-mez This is upmerged and ready for staging. 

I'll handle the last few comments you made before we merge this.

File: pkg/trace/writer/sender.go
Comment (knusbaum) [inline pkg/trace/writer/sender.go:225]: @ahmed-mez I'm going to leave this one undocumented for now. I'm not sure it's useful to customers yet. It'll mostly be useful for us until we understand what we're looking for.

File: pkg/trace/api/api.go
Comment (knusbaum) [reply pkg/trace/api/api.go:683]: https://github.com/DataDog/documentation/pull/19256

File: pkg/trace/writer/sender.go
Comment (knusbaum) [reply pkg/trace/writer/sender.go]: Added an undocumented config for this. We can remove it later.

**Comment by knusbaum** [comment]
@buraizu I could use another review on the release note. I've rewritten part of it.

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (knusbaum) [reply releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml]: > I don't think it's a big issue to allow these to be configurable, but maybe not as global environment variables? Wdyt?

Per our discussion, we will leave these in but undocumented for now until we see if customers need them at all, and how to best communicate what settings to adjust based on issues they have.

File: releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml
Comment (buraizu) [inline releasenotes/notes/trace-agent-pipeline-performance-fb64279908ceffce.yaml:21]:
```suggestion
    memory work at least as well, and often better, with these changes compared to previous Agent versions. This means users do not have to adjust
    their configuration to take advantage of these changes, and they do not
```

**Review by buraizu — APPROVED** [review]
Thanks for applying those changes, approving with a few additional edits per the style guide.

**Review by ahmed-mez — APPROVED** [review]

**Review by ogaca-dd — APPROVED** [review]

**Pull request merged**
